src/data/augment_layers.py

```diff
@@ -0,0 +1,600 @@
+        augment_selection.append(tf.keras.layers.Cropping2D(cropping=(0, 25)))
+
+    if args.aug_blur:
+        logger.info("Data augment: blur_image")
+        augment_selection.append(BlurImageLayer())
+
+    if args.aug_invert:
+        logger.info("Data augment: aug_invert")
+        augment_selection.append(InvertImageLayer(channels=args.channels))
+
+    return augment_selection
+
+
+def make_augment_model(augment_options, augment_selection=None):
+    """
+    Constructs an image augmentation model from a list of augmentation options,
+    with specific handling for certain layer combinations. If no
+    augment_selection is specified then a list of random augmentations are
+    generated.
+
+    Parameters
+    ----------
+    augment_options : list
+        List of possible augmentation layer instances.
+    augment_selection : list, optional
+        Preselected list of augmentation layers. If empty or None, a random
+        selection is made.
+
+    Returns
+    -------
+    tf.keras.Sequential
+        A TensorFlow Keras Sequential model comprising the selected
+        augmentation layers.
+
+    """
+
+    # Check if both types of binarization are present, remove Sauvola if true
+    otsu_present = any(isinstance(obj, BinarizeLayer)
+                       and obj.method == 'otsu' for obj in augment_selection)
+    sauvola_present = any(isinstance(obj, BinarizeLayer)
+                          and obj.method == 'sauvola' for obj in
+                          augment_selection)
+
+    if otsu_present and sauvola_present:
+        # Remove the object of type BinarizeLayer with Sauvola method
+        augment_selection = [obj for obj in augment_selection
+                             if not (isinstance(obj, BinarizeLayer)
+                                     and obj.method == 'sauvola')]
+
+    # Init blur params
+    mild_blur = False
+    blur_index = -1
+
+    # Check if a blur occurs before binarization, if so replace with mild blur
+    for i, augment in enumerate(augment_selection):
+        if isinstance(augment, BlurImageLayer):
+            mild_blur = True
+            blur_index = i
+        elif mild_blur and isinstance(augment, BinarizeLayer):
+            augment_selection[blur_index] = BlurImageLayer(mild_blur=True)
+
+    # Check if blur/distort occurs after binarization, if so move it in front
+    binarize = False
+    binarize_index = -1
+    for i, augment in enumerate(augment_selection):
+        if isinstance(augment, BinarizeLayer):
+            binarize = True
+            binarize_index = i
+        elif binarize and (isinstance(augment, BlurImageLayer)):
+            # Remove BlurImageLayer that occurs after binarization
+            augment_selection.remove(augment)
+            # Insert a mild blur object right before binarize
+            augment_selection.insert(binarize_index,
+                                     BlurImageLayer(mild_blur=True))
```

Comment:
Werkt dit? Lijkt me niet dat je iets mag inserten in een lijst waar je op dat moment doorheen aan het loopen bent. Als het wel kan, lijkt me alleen maar dat er nare side-effects van zullen komen.

Reply:
Misschien beter om de loop te doen met `for i, augment in enumerate(augment_selection[:]):`